### PR TITLE
Fix create_nnx_model() to load NNX checkpoint, verified with vllm_decode

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

- Update `create_nnx_model()` to load both Linen and NNX checkpoint
- FIXES: b/450116695

# Tests

Tested loading both linen and NNX checkpoint and compare the params before and after updating the model.

Loading linen checkpoint: https://paste.googleplex.com/5840546769797120

Loading NNX checkpoint (without pathways): https://paste.googleplex.com/5153322256433152

Loading NNX checkpoint (with pathways): https://paste.googleplex.com/5967825172824064

vLLM Decoding Results: http://shortn/_QHI6tVbeYP

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/model_creation_utils.py [hunk 1]
```diff
@@ -137,33 +137,64 @@ def create_sharded_state():
...
+          target_for_restore = jax.tree.map(
+              lambda v: v.value,
+              sharded_state,
+              is_leaf=lambda n: isinstance(n, nnx.Variable),
+          )
+
+          item_to_restore = {"params": {"params": target_for_restore}}
+          restore_args = {"params": {"params": ocp.checkpoint_utils.construct_restore_args(target_for_restore)}}
+        else:
+          # structure of nnx checkpoint: {'decoder': {'value': ...}}
```
Comment (hengtaoguo) [inline src/MaxText/model_creation_utils.py:174]: For your comments:
```
# structure of linen checkpoint: {'params': {'params': {'decoder': ...}}}
# structure of nnx checkpoint: {'decoder': {'value': ...}}
```

Is the `...` part structural identical between linen and nnx? Then for the below variable `checkpoint`, what's the tree structure of it?

Comment (SurbhiJainUSC) [reply src/MaxText/model_creation_utils.py:174]: Yes, `...` is the identical part. Here is what `checkpoint` variable would look like for both Linen and NNX: https://paste.googleplex.com/4702000029761536

File: src/MaxText/vllm_decode.py [hunk 2]
```diff
@@ -15,12 +15,12 @@
...
 Example command:
-  python3 -m MaxText.vllm_decode MaxText/configs/sft.yml \
+  python3 -m MaxText.vllm_decode MaxText/configs/base.yml \
     model_name=llama3.1-8b tokenizer_path=meta-llama/Llama-3.1-8B-Instruct \
     tokenizer_type=huggingface hf_access_token=<your_hf_token> \
     load_parameters_path=<your_checkpoint_path> \
-    per_device_batch_size=1 run_name=vllm_decode_test \
-    use_chat_template=True prompt="Suggest some famous landmarks in London." \
+    per_device_batch_size=1 run_name=vllm_decode_test max_target_length=64 \
+    use_chat_template=False prompt="Suggest some famous landmarks in London." \
```
Comment (A9isha) [inline src/MaxText/vllm_decode.py:23]: what is the issue if we use `sft.yml`?

File: src/MaxText/model_creation_utils.py [hunk 3]
```diff
@@ -137,33 +137,64 @@ def create_sharded_state():
...
         # waste memory, we instead restore the params field of the checkpoint (which itself may be a dictionary
         #  containing a key named 'params').
-        restore_args = ocp.checkpoint_utils.construct_restore_args(target_for_restore)
+
+        # Get the structure of checkpoint in `config.load_parameters_path`
+        metadata = ckptr.metadata(config.load_parameters_path)
+
+        is_nnx_checkpoint = True
+        if (
+            "params" in metadata.item_metadata.tree.keys()
```
Comment (A9isha) [inline src/MaxText/model_creation_utils.py:160]: I think the better way to decide if it is an NNX checkpoint is to check for nnx.Variable in the leaf.

File: src/MaxText/model_creation_utils.py [hunk 4]
```diff
@@ -137,33 +137,64 @@ def create_sharded_state():
...
+            and "params" in metadata.item_metadata.tree.get("params", {}).keys()
+        ):
+          # structure of linen checkpoint: {'params': {'params': {'decoder': ...}}}
+          is_nnx_checkpoint = False
+          target_for_restore = jax.tree.map(
+              lambda v: v.value,
+              sharded_state,
+              is_leaf=lambda n: isinstance(n, nnx.Variable),
```
Comment (A9isha) [inline src/MaxText/model_creation_utils.py]: if it is not an NNX checkpoint i.e., `is_nnx_checkpoint = False`, why do we have `nnx.Variable` check?

File: src/MaxText/vllm_decode.py (hunk 2 above)
Comment (SurbhiJainUSC) [reply src/MaxText/vllm_decode.py:23]: This `vllm_decode` script is not specifically for SFT. It can be used for decoding any NNX MaxText model. So, it doesn't make sense to use `sft.yml` configuration. Hence, I changed it to `base.yml`.

File: src/MaxText/model_creation_utils.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply src/MaxText/model_creation_utils.py:160]: This will check the params structure in the checkpoint present in GCS. Linen checkpoint has nested params structure but NNX checkpoint doesn't. So, this checks the structure of checkpoint in GCS before any loading begins.

File: src/MaxText/model_creation_utils.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply src/MaxText/model_creation_utils.py]: Updated the check.

File: src/MaxText/model_creation_utils.py (hunk 3 above)
Comment (A9isha) [reply src/MaxText/model_creation_utils.py:160]: Right, my comment was more to say that instead of relying on the existence of "params" in the keys, a better way to reliably infer that it is an NNX checkpoint is to check for the existence of `nnx.variable`.

But, it is okay for now.
